### PR TITLE
Add tests for active docs OAS3

### PR DESCRIPTION
**Pull request by JaurbanRH**

### CODE REVIEW

File: testsuite/tests/ui/oas/test_active_docs_v3.py [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+    }
+    oas_spec = resources.files("testsuite.resources.oas3").joinpath("petstore-expanded.json").read_text()
+    json_spec = json.loads(oas_spec)
+    json_spec['paths']['/pets']['get']['parameters'].append(user_key)
+    oas_spec = json.dumps(json_spec)
+    return oas_spec
+
+
+@pytest.fixture(scope="module")
+def prod_client(prod_client):
```
Comment (pehala) [inline testsuite/tests/ui/oas/test_active_docs_v3.py:55]: If you use prod_client you need to mark this test as disruptive. Why do you need production gateway in the first place?

Comment (JaurbanRH) [reply testsuite/tests/ui/oas/test_active_docs_v3.py:55]: Because oas3 active docs make request against production endpoint of service.

Comment (pehala) [reply testsuite/tests/ui/oas/test_active_docs_v3.py:55]: Ok, then you can either mark this test as disruptive or run it with self-managed APIcast

File: testsuite/ui/widgets/__init__.py [hunk 2]
```diff
@@ -336,6 +336,57 @@ def get_response_code(self):
...
+# Widget contains fill method which raise not implemented exception if widget is not fillable but pylint detect it as
+# an abstract method
+class ActiveDocV3Section(Widget):
+    """Active Doc V3 preview section"""
+    ROOT = ParametrizedLocator('//*[@class="operation-tag-content"]')
+    ENDPOINTS_LIST = './span/div'
+    ITEMS_LOCATOR = './span/div/div/button/span[2]/a/span'
+    ITEMS_EXPAND_LOCATOR = './span/div/div/button/span[text()="{}"]/../span[2]/a/span[contains(text(),"{}")]'
+    RESPONSE_CODE_LOCATOR = './/*[text()="Server response"]/../table/tbody/tr/td[1]'
+    TRY_OUT_BUTTON_LOCATOR = '//*[@class="btn try-out__btn"]'
```
Comment (jsmolar) [inline testsuite/ui/widgets/__init__.py:349]: Keep using dot notation when ROOT is defined

File: testsuite/ui/widgets/__init__.py [hunk 3]
```diff
@@ -336,6 +336,57 @@ def get_response_code(self):
...
+    SELECT_OPTION_LOCATOR = '//select/option[text()="{}"]'
+    EXECUTE_BUTTON_LOCATOR = '//*[@class="btn execute opblock-control__btn"]'
+
+    def __init__(self, parent=None, logger=None):
+        Widget.__init__(self, parent, logger=logger)
+
+    @property
+    def endpoints(self):
```
Comment (jsmolar) [inline testsuite/ui/widgets/__init__.py:357]: this method is defined as abstract in `ParametrizedView`

File: testsuite/ui/widgets/__init__.py [hunk 4]
```diff
@@ -336,6 +336,57 @@ def get_response_code(self):
         return self.browser.text(self.RESPONSE_CODE_LOCATOR, parent=self)
 
 
+# pylint: disable=abstract-method
+# Widget contains fill method which raise not implemented exception if widget is not fillable but pylint detect it as
+# an abstract method
```
Comment (jsmolar) [inline testsuite/ui/widgets/__init__.py:341]: You don't need to disable abstract-method.

File: testsuite/ui/widgets/__init__.py [hunk 5]
```diff
@@ -336,6 +336,57 @@ def get_response_code(self):
...
+        :param method string eg. GET, POST
+        :param key string name of application
+        """
+        self.browser.click(self.item_element(method, path))
+        self.browser.click(self.browser.element(self.TRY_OUT_BUTTON_LOCATOR))
+        self.browser.selenium.find_element_by_xpath(self.SELECT_OPTION_LOCATOR.format(key)).click()
+        self.browser.click(self.browser.element(self.EXECUTE_BUTTON_LOCATOR))
+
+    @backoff.on_exception(backoff.fibo, NoSuchElementException, max_tries=4, jitter=None)
+    def get_response_code(self):
```
Comment (jsmolar) [inline testsuite/ui/widgets/__init__.py:385]: Please look at `APIDocsView` and try to use the naming that is used there.

File: testsuite/ui/views/admin/audience/developer_portal/__init__.py [hunk 6]
```diff
@@ -72,7 +72,7 @@ def is_displayed(self):
 class ActiveDocsNewView(BaseAudienceView):
     """View representation of New Active Docs page"""
     path_pattern = '/admin/api_docs/services/new'
-    create_spec_btn = ThreescaleCreateButton()
+    create_spec_btn = GenericLocatorWidget("//button[contains(text(), 'Create')]")
```
Comment (jsmolar) [inline testsuite/ui/views/admin/audience/developer_portal/__init__.py:75]: Why this change?

File: testsuite/tests/ui/oas/test_active_docs_v3.py [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+                   "https://issues.redhat.com/browse/THREESCALE-1460")
+def test_active_docs_v3_generate_endpoints(login, navigator, service, prod_client, application, request, oas3_spec):
+    """
+    Test:
+        - Create service via API
+        - Create Active doc via UI
+        - Assert endpoints are correctly generated from oas specification
+        - Assert that try endpoint works as expected
+    """
+    name = blame(request, "active_doc_v3")
```
Comment (jsmolar) [inline testsuite/tests/ui/oas/test_active_docs_v3.py:77]: Spec creation should be fixture

**Review by jsmolar — CHANGES_REQUESTED** [review]

File: testsuite/ui/views/admin/audience/developer_portal/__init__.py (hunk 6 above)
Comment (JaurbanRH) [reply testsuite/ui/views/admin/audience/developer_portal/__init__.py:75]: Oas3 json contains word Create so we need to specify create button locator that it contains //button instead of //*

File: testsuite/ui/widgets/__init__.py (hunk 3 above)
Comment (JaurbanRH) [reply testsuite/ui/widgets/__init__.py:354]: After discussion with @dhlavac  we agreed that these changes (this one and next two) are unnecessary

File: testsuite/tests/ui/oas/test_active_docs_v3.py [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+                     service=service,
+                     oas_spec=oas3_spec,
+                     publish_option=True,
+                     skip_validation_option=True)
+    return service.active_docs.list()[0]
+
+
+# pylint: disable=unused-argument, too-many-arguments
+@pytest.mark.disruptive  # test should be mark as disruptive because of production gateway redeploy
+@pytest.mark.issue("https://issues.redhat.com/browse/THREESCALE-770",
```
Comment (dhlavac) [inline testsuite/tests/ui/oas/test_active_docs_v3.py:85]: Issue 770 includes a service discovery feature. It should be covered with the test or removed from this test.

**Review by dhlavac — APPROVED** [review]

**Review by pehala — APPROVED** [review]

**Review by jsmolar — APPROVED** [review]
